contracts/tokenwrapper/src/contract.rs

```diff
@@ -0,0 +1,533 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    attr, coins, from_binary, to_binary, Addr, BankMsg, Binary, CosmosMsg, Deps, DepsMut, Env,
+    MessageInfo, Response, StdError, StdResult, Uint128, WasmMsg,
+};
+use cw2::set_contract_version;
+
+use cw20::{BalanceResponse, Cw20ExecuteMsg, Cw20QueryMsg, Cw20ReceiveMsg, TokenInfoResponse};
+use cw20_base::allowances::{
+    execute_burn_from, execute_decrease_allowance, execute_increase_allowance, execute_send_from,
+    execute_transfer_from, query_allowance,
+};
+use cw20_base::contract::{
+    execute_burn, execute_mint, execute_send, execute_transfer, query_balance, query_token_info,
+};
+
+use cw20_base::state::{MinterData, TokenInfo, TOKEN_INFO};
+
+use protocol_cosmwasm::error::ContractError;
+use protocol_cosmwasm::token_wrapper::{Cw20HookMsg, ExecuteMsg, InstantiateMsg, QueryMsg};
+
+use crate::state::{Supply, TOTAL_SUPPLY};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cosmwasm-tokenwrapper";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // store token info using cw20-base format
+    let data = TokenInfo {
+        name: msg.name,
+        symbol: msg.symbol,
+        decimals: msg.decimals,
+        total_supply: Uint128::zero(),
+        // set self as minter, so we can properly execute mint and burn
+        mint: Some(MinterData {
+            minter: env.contract.address,
+            cap: None,
+        }),
+    };
+    TOKEN_INFO.save(deps.storage, &data)?;
+
+    // set supply to 0
+    let supply = Supply::default();
+    TOTAL_SUPPLY.save(deps.storage, &supply)?;
+
+    Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Wrap {} => wrap_native(deps, env, info),
```

Comment:
Why is there no wrap another CW20 token and only native? Is this intentional?

Reply:
https://github.com/webb-tools/protocol-cosmwasm/blob/guorong/implement-tokenwrapper/contracts/tokenwrapper/src/contract.rs#L74
This is where the cw20 wrap is done.
The reason of implementing the cw20 wrapping in the `ReceiveMsg` is that it comes from the characteristic of cosmwasm.
In cosmwasm(cw20), the delegated call is performed with this `Send/Receive` interface.
If you want more detail, you can check https://docs.cosmwasm.com/cw-plus/0.9.0/cw20/spec